### DISCUSSION TITLE
SystemSet with PhantomData, or ordering system by data dependencies

### DISCUSSION THREAD

**Discussion by hyranno** [bevyengine/bevy#10155] [Q&A]
Hello.
I'm trying to determine system ordering by data dependencies. What I want is like this.
The system produce the subjects by the dependencies.
```Rust
pub fn add_data_dependency<Marker>(
    app: &mut App,
    system: impl IntoSystemConfigs<Marker>,
    subjects: Vec<Box<dyn SystemSet>>,
    dependencies: Vec<Box<dyn SystemSet>>,
) {
    /* ... */
}
```

So I wrote the code below, to represent `SystemSet` which produces data `T`.
But this doesn't compile, saying `T` needs to be `Clone`, `Eq`, `Send`, ... and so on.
```Rust
use std::marker::PhantomData;

use bevy::prelude::*;

#[derive(SystemSet, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct DataLabel<T: 'static> {
    _phantom: PhantomData<T>,
}
impl<T: 'static> DataLabel<T> {
    pub fn set() -> Box<dyn SystemSet> {
        Box::new( Self{_phantom: PhantomData } )
    }
}
```

I found Rust issue which may relate (https://github.com/rust-lang/rust/issues/26925), but not understand well.

Is there a way to make it compile?
Or, existing way to ordering system by data dependencies?

### more explanation
For example,  consider `system1` write `ComponentA` and `system2` read `ComponentA`.
I say `system2` depends `ComponentA`.
In standard form, we can `system2.after(system1)`.
But I want to write `system2.after(DataLabel<ComponentA>)`, where `DataLabel<ComponentA>` is `SystemSet` which may write to `ComponentA`. I mean this for ordering system by data dependencies.
In this form, we don't need to list all systems which may write to `ComponentA`. We can just think about the `system2`.
I'm trying to write function `add_data_dependencies` which add system to `DataLabel<Subject>` and order them after `DataLabel<Dependency>`.
Then the problem, I can't write `DataLabel<T>`, the `SystemSet` with `PhantomData<T>`.

**Comment by nicopap** [comment]
I don't understand what you mean by "ordering by data dependency". You seem to want to have just a way to order systems? Check out the following example: <https://github.com/bevyengine/bevy/blob/v0.11.3/examples/ecs/ecs_guide.rs#L293-L315>

You have two ways to order systems:

1. Using `.after(foo)` and `.before(bar)` combinators

```rust
app.add_systems(Update, (first_system.before(second_system), second_system));
```

2. Using `.chain()` 

```rust
app.add_systems(Update, (first_system, second_system).chain());
```

Within a single `.chain()`, you can also run systems in parallel as follow

```rust
app.add_systems(Update, (
    (system1a, system1b, system1c),
    system2a,
    (system3a, system3b),
).chain());
```


Maybe you don't mean that? But I really don't understand what "Ordering by data dependency" means. Fundamentally, even if two systems access the same data (for example, a resource or component), you still need to specify an order. It's not enough to know that two systems access the same data, hence, it's not possible to "order by data dependency". But I don't think that's what you mean.

**Reply by hyranno** [reply]
Thank you for reply, and sorry for lack of explanation.
I added explanation for what I'm trying to do.

**Comment by jancespivo** [comment]
Hi, it might be completely unrelated but I struggled with `PhantomData` as well until I found this recipe:

```rust
#[derive(Default)]
pub struct MyStruct<T> {
    marker: PhantomData<fn() -> T>
}
```

And also I would start with the minimal solution so when you want to derive SystemSet and set bound T to Component

```rust
#[derive(Default, SystemSet)]
pub struct DataLabel<T: Component> {
    _phantom: PhantomData<fn() -> T>,
}
```

I hope I've helped at least a bit ;) I'm quite curious what you want to achieve.

**Reply by hyranno** [reply]
Unfortunately, that doesn't work.
`impl` thing doesn't compile yet, saying trait bounds on T are not satisfied.

**Comment by hyranno** [comment]
Well, I decided to implement traits by hand without derive macro.
And what I wanted to do (ordering system by data dependencies) is now become a tiny crate : https://github.com/hyranno/bevy_tiny_data_flow .

Then, now I found https://github.com/bevyengine/bevy/pull/8595 .
Looks like better.

Any way, thank you!
